module/zfs/zfs_ioctl.c

```diff
@@ -2819,8 +2777,49 @@ zfs_ioc_inherit_prop(zfs_cmd_t *zc)
 			return (SET_ERROR(EINVAL));
 	}
 
-	/* property name has been validated by zfs_secpolicy_inherit_prop() */
-	return (dsl_prop_inherit(zc->zc_name, zc->zc_value, source));
+	if (prop == ZPROP_INVAL) {
+		if (!zfs_prop_user(propname))
+			return (SET_ERROR(EINVAL));
+
+		type = PROP_TYPE_STRING;
+	} else if (prop == ZFS_PROP_VOLSIZE || prop == ZFS_PROP_VERSION) {
+		return (SET_ERROR(EINVAL));
+	} else {
+		type = zfs_prop_get_type(prop);
+	}
+
+	/*
+	 * zfs_prop_set_special() expects properties in the form of an
+	 * nvpair with type info.
+	 */
+	dummy = fnvlist_alloc();
+
+	switch (type) {
+	case PROP_TYPE_STRING:
+		VERIFY(0 == nvlist_add_string(dummy, propname, ""));
+		break;
+	case PROP_TYPE_NUMBER:
+	case PROP_TYPE_INDEX:
+		VERIFY(0 == nvlist_add_uint64(dummy, propname, 0));
+		break;
+	default:
+		err = SET_ERROR(EINVAL);
+		goto errout;
+	}
+
+	pair = nvlist_next_nvpair(dummy, NULL);
+	if (pair == NULL) {
+		err = SET_ERROR(EINVAL);
+	} else {
+		err = zfs_prop_set_special(zc->zc_name, source, pair);
+		if (err == -1) /* property is not "special", needs handling */
+			err = dsl_prop_inherit(zc->zc_name, zc->zc_value,
```

Comment:
>would it make sense to set received == TRUE unconditionally in the user-space for the snapdev only ?

No, `received` serves a totally different purpose and we need both TRUE and FALSE, we cannot set `received == TRUE` unconditionally from userland for snapdev.

The intention is to call `zfs_prop_set_special()` for every property, not just `snapdev`. This seems only natural and is just a coincidence we haven't needed this until now, just because snapdev is the only "special" inheritable property: for this reason i don't like adding a comment that would specify the obvious case ("_call zfs_prop_set_special() for every property, not just "snapdev"_") because it could be confusing to anyone reading the code.

Now, if someone wanted to know why it was implemented this way, when this was changed, and maybe how it was working before, they can read the VCS history which explains our reason to make this change in the commit message.